### PR TITLE
[Client] Use async rpc for remote call and actor creation

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
Currently in Ray client, each call to schedule a task, an actor or an actor method requires the client to wait synchronously for object or actor ID(s) to return from the cluster. Running the following expression has been slow when there is non-negligible latency between Ray client and cluster: `refs = [f.remote() for _ in range(1000)]`. Each `f.remote()` call blocks on a client-cluster round trip.

This change allows `ClientObjectRef` and `ClientActorRef` to be created with a `Future` instead of a `bytes` ID, making the RPC response from `Schedule` calls non-blocking. Underlying object and actor IDs need only to be waited for when these references are used.

Additional changes:
- Scheduling calls are moved to use streaming RPC, together with other operations like `get` and `put`. So task and method calls are still serialized.
- For name actors created from `get_actor()`, they lack metadata such as per method signature and `num_returns`. Make a separate call to retrieve `num_returns` and method signatures.
- Check method signatures at client, so callers can still get function signature mismatch errors synchronously. This behavior is the same as when running as a driver.

Performance is tested by running Ray client from a laptop, and a Ray cluster with 1 `m5.2xlarge` node in AWS `us-west-2`.  Before the change,
| Measurements | Time in seconds |
| --------------- | ---------------- |
| 40 remote actor method calls (walltime in script) | 1.437 |
| 40 remote actor method calls + get 40 ObjectRefs in a list (walltime in script) | 1.488 |
Time to start calls (from before making actor method calls, to method called) | max: 1.4865, min: 0.0829, median: 0.7895 |

After the change,
| Measurements | Time in seconds |
| --------------- | ---------------- |
| 40 remote actor method calls (walltime in script) | 0.004 |
| 40 remote actor method calls + get 40 ObjectRefs in a list (walltime in script) | 0.095 |
Time to start calls (from before making actor method calls, to method called) | max: 0.1090, min: 0.0863, median: 0.0987 |

Round trip latency between client and cluster is mostly hidden from remote calls now.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by mwtian** [comment]
cc @iycheng

**Comment by AmeerHajAli** [comment]
Wow @mwtian this looks really good. Do you have early performance numbers?

File: python/ray/includes/object_ref.pxi [hunk 1]
```diff
@@ -88,8 +88,8 @@ cdef class ObjectRef(BaseID):
     def call_site(self):
         return decode(self.call_site_data)
 
-    cdef size_t hash(self):
-        return self.data.Hash()
+    def size(self):
```
Comment (ijrsvt) [inline python/ray/includes/object_ref.pxi:91]: can you revert these changes?

File: python/ray/includes/unique_ids.pxi [hunk 2]
```diff
@@ -323,10 +328,40 @@ cdef class ClientActorRef(ActorID):
...
+
     @property
     def id(self):
         return self.binary()
 
+    cdef _set_id(self, id):
+        check_id(id, CActorID.Size())
+        self.data = CActorID.FromBinary(<c_string>id)
+        client.ray.call_retain(id)
+
```
Comment (ijrsvt) [inline python/ray/includes/unique_ids.pxi:359]: Could you add some tests to ensure that the 'public API' for ClientObjects and non-client Objects are the same?

File: python/ray/util/client/worker.py [hunk 3]
```diff
@@ -313,31 +314,39 @@ def wait(self,
 
         return (client_ready_object_ids, client_remaining_object_ids)
 
-    def call_remote(self, instance, *args, **kwargs) -> List[bytes]:
+    def call_remote(self, instance, *args, **kwargs) -> List[Future]:
```
Comment (ijrsvt) [inline python/ray/util/client/worker.py:317]: Can you make sure that `call_remote` in api.py also has the updated type? I think it might also be good to change `call_remote` in api.py to `_call_remote` to make it clear that it is not for public usage.

**Review by ijrsvt — COMMENTED** [review]
Looks pretty good overall! Just some initial feedback!

**Comment by mwtian** [comment]
> Wow @mwtian this looks really good. Do you have early performance numbers?

Microbenchmarks do show a lot of difference. Trying to run it in a cluster environment. Will update when the numbers are available.

File: python/ray/util/client/worker.py (hunk 3 above)
Comment (ijrsvt) [reply python/ray/util/client/worker.py:318]: A TODO is totally fine! I think moving all of those to internal would be good in a follow up :)

File: python/ray/includes/object_ref.pxi (hunk 1 above)
Comment (ijrsvt) [reply python/ray/includes/object_ref.pxi:91]: Ahh okay!

**Comment by mwtian** [comment]
> Wow @mwtian this looks really good. Do you have early performance numbers?

Updated latency stats in the PR description. Basically most of client-cluster round trip latency is hidden from remote calls now.

File: python/ray/tests/BUILD [hunk 4]
```diff
@@ -316,7 +316,7 @@ py_test_module_list(
     "test_multiprocessing.py",
     "test_list_actors.py",
   ],
-  size = "medium",
+  size = "large",
```
Comment (ijrsvt) [inline python/ray/tests/BUILD:319]: Did this test start taking longer because of this PR?

File: python/ray/tests/test_basic_2.py [hunk 5]
```diff
@@ -60,9 +60,9 @@ def f2(x, y, *args):
     assert ray.get(f1.remote()) == ()
     assert ray.get(f1.remote(1)) == (1, )
     assert ray.get(f1.remote(1, 2, 3)) == (1, 2, 3)
-    with pytest.raises(Exception):
+    with pytest.raises(TypeError):
```
Comment (ijrsvt) [inline python/ray/tests/test_basic_2.py:63]: Are these changes necessary for the test to work, or are they just making the tests more specific (better)

File: python/ray/util/client/worker.py [hunk 6]
```diff
@@ -500,14 +514,12 @@ def internal_kv_list(self, prefix: bytes) -> bytes:
...
-        return json.loads(
-            self.server.ListNamedActors(req,
-                                        metadata=self.metadata).actors_json)
+        return json.loads(self.data_client.ListNamedActors(req).actors_json)
 
     def is_initialized(self) -> bool:
         if self.server is not None:
             return self.get_cluster_info(
-                ray_client_pb2.ClusterInfoType.IS_INITIALIZED)
+                ray_client_pb2.ClusterInfoType.IS_INITIALIZED) is not None
```
Comment (ijrsvt) [inline python/ray/util/client/worker.py:522]: I think the return value from the server is a boolean (T/F). I think we want to keep directly returning it.

File: python/ray/tests/BUILD (hunk 4 above)
Comment (mwtian) [reply python/ray/tests/BUILD:319]: There seems to be an issue with test_basic_2.py test_actor_recursive, but I'm not able to reproduce it locally. Still investigating.

File: python/ray/tests/test_basic_2.py (hunk 5 above)
Comment (mwtian) [reply python/ray/tests/test_basic_2.py:63]: Making the tests more specific for both client and driver modes.

Comment (ijrsvt) [reply python/ray/tests/test_basic_2.py:63]: Fantastic :) This is definitely helpful!

File: python/ray/util/client/worker.py (hunk 6 above)
Comment (mwtian) [reply python/ray/util/client/worker.py]: Reverted.

**Comment by mwtian** [comment]
Rebased with on master branch. Ready for reviews now.

File: python/ray/tests/test_asyncio.py [hunk 7]
```diff
@@ -304,15 +304,22 @@ def interceptor(e):
     ray.worker._unhandled_error_handler = interceptor
     x1 = f.remote()
     del x1
-    wait_for_condition(lambda: num_exceptions == 1)
+    if client_test_enabled():
+        # In client mode, x1 may not have received the underlying exception
+        # when deleted, so no exception is raised.
+        prev_num_exceptions = num_exceptions
```
Comment (AmeerHajAli) [inline python/ray/tests/test_asyncio.py:310]: is this a bug? Should it behave like native ray?

File: python/ray/tests/test_asyncio.py [hunk 8]
```diff
@@ -304,15 +304,22 @@ def interceptor(e):
...
+        prev_num_exceptions = 1
 
     # Test we don't report handled exceptions.
     x1 = f.remote()
     with pytest.raises(ray.exceptions.RayError):
         await x1
     del x1
     await asyncio.sleep(1)
-    assert num_exceptions == 1, num_exceptions
+    assert num_exceptions == prev_num_exceptions, (num_exceptions,
```
Comment (AmeerHajAli) [inline python/ray/tests/test_asyncio.py:321]: maybe explicitly write expected number of exceptions: num_exceptions received number of exceptions: prev_num_exceptions in the error message

File: python/ray/tests/test_client.py [hunk 9]
```diff
@@ -135,14 +135,21 @@ def f():
...
         refs = [f.remote() for _ in range(100)]
-        with ray.worker.data_client.lock:
-            req_id_before = ray.worker.data_client._req_id
         assert ray.get(refs) == ["OK" for _ in range(100)]
+
         # Only 1 RPC should be sent.
-        with ray.worker.data_client.lock:
-            assert ray.worker.data_client._req_id == req_id_before + 1, \
-                ray.worker.data_client._req_id
+        assert get_count == 1
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client.py:152]: I don't understand this change. Can you please explain?

File: python/ray/tests/test_client.py [hunk 10]
```diff
@@ -331,16 +338,27 @@ def __init__(self):
 
             def say_hello(self, whom):
                 self.count += 1
-                return ("Hello " + whom, self.count)
+                return "Hello " + whom, self.count
+
+            @ray.method(num_returns=2)
+            def say_hi(self, whom):
+                self.count += 1
+                return "Hi " + whom, self.count
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client.py:346]: why this is needed?

File: python/ray/tests/test_client.py [hunk 11]
```diff
@@ -331,16 +338,27 @@ def __init__(self):
...
-        s, count = ray.get(actor.say_hello.remote("world"))
+
+        ref = actor.say_hello.remote("world")
+        s, count = ray.get(ref)
         assert s == "Hello world"
         assert count == 2
 
+        r1, r2 = actor.say_hi.remote("ray")
+        assert ray.get(r1) == "Hi ray"
+        assert ray.get(r2) == 3
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client.py:360]: Is this to improve the test? Shouldn't it behave like it was before making this change?

Comment (AmeerHajAli) [reply python/ray/tests/test_client.py:360]: (just double checking to make sure the async behavior did not introduce an issue)

File: python/ray/tests/test_asyncio.py (hunk 7 above)
Comment (mwtian) [reply python/ray/tests/test_asyncio.py:310]: I think not really. Exception handling and behavior during {Client}ObjectRef destruction is not a public API. Client uses async communication with cluster, but scripts running in cluster can communicate with workers synchronously.

File: python/ray/tests/test_client_terminate.py [hunk 12]
```diff
@@ -113,11 +113,12 @@ def f():
         class MetadataIsCorrectlyPassedException(Exception):
             pass
 
-        def mock_terminate(term, metadata):
-            raise MetadataIsCorrectlyPassedException(metadata[1][0])
+        def mock_terminate(self, term):
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_terminate.py:116]: what term is used for?

File: python/ray/tests/test_client_warnings.py [hunk 13]
```diff
@@ -19,33 +17,6 @@ def reset_debug_logs():
...
-            def f():
-                return 42
-
-            with warnings.catch_warnings(record=True) as warn_list:
-                for _ in range(TASK_WARNING_THRESHOLD):
-                    f.remote()
-            assert not any(f"More than {TASK_WARNING_THRESHOLD} remote tasks "
-                           "have been scheduled." in str(w.args[0])
-                           for w in warn_list)
-
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_warnings.py:48]: Why are we removing this? this is one of our important usability warnings.

File: python/ray/tests/test_client.py (hunk 9 above)
Comment (mwtian) [reply python/ray/tests/test_client.py:152]: Basically, `ray.get()` is converted to use unary-unary RPC instead of streaming RPC, so we need another way to check if only 1 RPC request is sent. `ray.get()` does not need to serialize with other operations. In future it will be converted to unary-streaming to chunk response data.

File: python/ray/tests/test_client.py (hunk 10 above)
Comment (mwtian) [reply python/ray/tests/test_client.py:346]: Test `@ray.method(num_returns=2)` is handled correctly.

File: python/ray/tests/test_client_warnings.py (hunk 13 above)
Comment (AmeerHajAli) [reply python/ray/tests/test_client_warnings.py:48]: CC @ckw017

File: python/ray/tests/test_client.py (hunk 11 above)
Comment (mwtian) [reply python/ray/tests/test_client.py:360]: Yes, this is an enhancement to ensure the behavior is the same between cluster and client modes, for `@ray.method(num_returns=2)`.

File: python/ray/tests/test_client_terminate.py (hunk 12 above)
Comment (mwtian) [reply python/ray/tests/test_client_terminate.py:116]: This mocks the `Terminate()` handler, which accepts the `TerminateRequest` (`term`). We need the same function signature.

File: python/ray/tests/test_client_warnings.py (hunk 13 above)
Comment (mwtian) [reply python/ray/tests/test_client_warnings.py:48]: Yes, because it is no longer necessary with async task calls.

Comment (AmeerHajAli) [reply python/ray/tests/test_client_warnings.py:48]: Wouldn't the overall performance be the same if the network BW is abused?

Comment (mwtian) [reply python/ray/tests/test_client_warnings.py:48]: Each task is very small e.g. < 1kb. The issue has been blocking on round trip latency instead of throughput, which is resolved now. And in general, having unnecessary warnings can confuse users and degrade the experience.

Comment (AmeerHajAli) [reply python/ray/tests/test_client_warnings.py:48]: Should we also remove the logic this is testing?

**Comment by pcmoritz** [comment]
There is some complexity introduced here since the client now has to know both the future AND the object reference (the latter will become available later). Is it necessary for the client to know the object reference? I wonder if the cleaner design would be to have only a client side ID/future for every .remote call on the client side and always resolve that to the underlying object reference on the server side. This will get rid of the awkward mutexes introduced in this PR.

File: python/ray/tests/test_client.py [hunk 14]
```diff
@@ -454,8 +472,11 @@ def print_on_stderr_and_stdout(s):
 @pytest.mark.skipif(sys.platform == "win32", reason="Failing on Windows.")
 def test_serializing_exceptions(ray_start_regular_shared):
     with ray_start_client_server() as ray:
-        with pytest.raises(ValueError):
-            ray.get_actor("abc")
+        with pytest.raises(
+                ValueError, match="Failed to look up actor with name 'abc'"):
+            a = ray.get_actor("abc")
+            # Reading underlying ActorID to raise the Exception from server.
+            print(a.id)
```
Comment (pcmoritz) [inline python/ray/tests/test_client.py:479]: This really shouldn't be needed, please make the test pass without this!

**Comment by mwtian** [comment]
> There is some complexity introduced here since the client now has to know both the future AND the object reference (the latter will become available later). Is it necessary for the client to know the object reference? I wonder if the cleaner design would be to have only a client side ID/future for every .remote call on the client side and always resolve that to the underlying object reference on the server side. This will get rid of the awkward mutexes introduced in this PR.

I thought about this approach at the beginning. One complexity / inefficiency is that if the user uses the ClientObjectRef / ClientActorRef as task args or captures the references in tasks / actors, we need to run another lookup to find the corresponding ref in the cluster. Another the downside is that if some users managed to use ClientObjectRef / ClientActorRef out-of-band, i.e. using raw ids, switching to client-side ref would break them too.

File: python/ray/tests/test_client.py (hunk 14 above)
Comment (mwtian) [reply python/ray/tests/test_client.py:479]: Merge issue that left an obsolete logic. Will be removed!

File: python/ray/tests/test_client_references.py [hunk 15]
```diff
@@ -261,7 +285,8 @@ def check(self):
         def connect_api():
             api = _ClientContext()
             api.connect("localhost:50051", namespace="default_test_namespace")
-            api.get_actor("actor")
+            # Make a call to resolve the underlying ActorID.
```
Comment (pcmoritz) [inline python/ray/tests/test_client_references.py:288]: Again this shouldn't be needed

Comment (mwtian) [reply python/ray/tests/test_client_references.py:288]: Will revert.

File: python/ray/util/client/common.py [hunk 16]
```diff
@@ -217,61 +247,89 @@ def __init__(self,
         self.actor_ref = actor_ref
         self._dir: Optional[List[str]] = None
         if actor_class is not None:
-            self._dir = list(
-                dict(
-                    inspect.getmembers(actor_class.actor_cls,
-                                       is_function_or_method)).keys())
+            self._method_num_returns = {}
```
Comment (pcmoritz) [inline python/ray/util/client/common.py:250]: I suggest splitting this PR into two: One for the async submission, and one (smaller one) for the num_returns fix and the `data = ray.worker._dumps_from_client(self.actor_cls)` change above.

Maybe you have ideas how it could be split even more.

It is quite possible that this PR will cause problems further down the line (just by the nature of the change), so having it chunked into smaller pieces will make it much easier to debug problems.

Comment (mwtian) [reply python/ray/util/client/common.py:250]: Good suggestion, since we need to wait for other changes to merge too. I'm thinking of a preliminary change that allows 
1.  Client{Object,Actor}Ref to take futures
2. Store num_returns & function signatures, check them synchronously
(2) is needed by the async change for the (desirable) behavior of raise `TypeError` synchronously when functions are called with incorrect arguments. They should have no / minimal conflict with other PRs.

Comment (mwtian) [reply python/ray/util/client/common.py:250]: Splitted out a (majority) part of the code to #18677, which should not have significant behavior change.

**Comment by ericl** [comment]
Is this really a release blocker? It looks more like a performance optimization than a bug.

**Comment by mwtian** [comment]
> Is this really a release blocker? It looks more like a performance optimization than a bug.

It seems to be a significant usability issue frequently complained by Ray client users, although there are workarounds. It is ok to not block the release on this, if it does not meet the bar.

File: python/ray/tests/test_client_warnings.py (hunk 13 above)
Comment (ckw017) [reply python/ray/tests/test_client_warnings.py:48]: Sounds good to me, avoiding the round-trip cost should definitely be a big help for this case.

**Review by ckw017 — APPROVED** [review]
Looks good to me. Something nice to have (if this doesn't already exist) would be some tests to check if/how error propagation works over the datastream (i.e. ClientTaskTicket.valid is False).

File: python/ray/tests/test_placement_group_2.py [hunk 17]
```diff
@@ -46,28 +46,18 @@ def value(self):
...
-        except ValueError:
-            error_count = error_count + 1
-        assert error_count == 2
 
-        try:
+        with pytest.raises(ValueError, match="bundle index must be -1"):
             Actor.options(placement_group_bundle_index=0).remote()
-        except ValueError:
-            error_count = error_count + 1
-        assert error_count == 3
```
Comment (AmeerHajAli) [inline python/ray/tests/test_placement_group_2.py:61]: is this some unrelated fixes/improvements to the test?

**Review by AmeerHajAli — COMMENTED** [review]
@mwtian , can we please add more unit tests to the different stubs and in general?
Can we also please add the benchmarks or this is already in the existing benchmarks and we should just make sure it does not regress somehow?

**Review by ijrsvt — COMMENTED** [review]
I'm a bit concerned with the test changes. What happens without them?

File: python/ray/tests/test_placement_group_2.py (hunk 17 above)
Comment (mwtian) [reply python/ray/tests/test_placement_group_2.py:61]: Yes, this is the make the test more debuggable.

File: python/ray/includes/object_ref.pxi [hunk 18]
```diff
@@ -166,8 +166,19 @@ cdef class ClientObjectRef(ObjectRef):
...
+            try:
+                self._wait_for_id(timeout=10)
+            # cython would suppress this exception as well, but it tries to
+            # print out the exception which may crash. Log a simpler message
+            # instead.
+            except Exception:
+                logger.info(
+                    "Exception in ObjectRef is ignored in destructor. "
+                    "To raise the exception, use the object reference before "
+                    "its destructor runs.")
```
[the remark this answers is not in the document]
Comment (mwtian) [reply python/ray/includes/object_ref.pxi:179]: Exceptions cannot be raised from destructors and trying to print them can crash too. If the exception is a concern to users, it can be acted upon by calling `ray.get()` on the object ref before the object ref is destroyed, to raise the exception. Is there a better way to reword this?

File: python/ray/util/client/worker.py [hunk 19]
```diff
@@ -138,9 +136,7 @@ def __init__(
 
         self.closed = False
 
-        # Track these values to raise a warning if many tasks are being
-        # scheduled
-        self.total_num_tasks_scheduled = 0
+        # Track these values to raise a warning if too data are transferred.
```
Comment (iycheng) [inline python/ray/util/client/worker.py:139]: # Track these values to raise a warning if a lot of data is transferred.

Comment (iycheng) [inline python/ray/util/client/worker.py:139]: # Track these values to raise a warning if a lot of data is transferred.

File: python/ray/includes/unique_ids.pxi [hunk 20]
```diff
@@ -326,8 +329,19 @@ cdef class ClientActorRef(ActorID):
             # call_release in this case, since the client should have already
             # disconnected at this point.
             return
-        if client.ray.is_connected() and not self.data.IsNil():
-            client.ray.call_release(self.id)
+        if client.ray.is_connected():
+            try:
+                self._wait_for_id(timeout=10)
```
Comment (iycheng) [inline python/ray/includes/unique_ids.pxi:334]: The question here: any reason you pick 10s? Another question is that if it failed to be resolved, will it lead to a leak?

File: src/ray/protobuf/ray_client.proto [hunk 21]
```diff
@@ -367,6 +367,9 @@ message DataRequest {
     PrepRuntimeEnvRequest prep_runtime_env = 7;
     ConnectionCleanupRequest connection_cleanup = 8;
     AcknowledgeRequest acknowledge = 9;
+    ClientTask task = 10;
+    TerminateRequest terminate = 11;
+    ClientListNamedActorsRequest list_named_actors = 12;
```
Comment (iycheng) [inline src/ray/protobuf/ray_client.proto:372]: It looks like we move some functions to the data client. Should we clean up the raylet one, like Schedule/ListNamedActors ?

**Review by fishbone — COMMENTED** [review]
Sorry for being late! Overall it looks clean, I only have one big concern about the leak.

Another thing is that should we clean up the old code? It looks like we've moved some functions to the dataclient.

**Comment by iycheng** [comment]
> > @mwtian , can we please add more unit tests to the different stubs and in general?
> > Can we also please add the benchmarks or this is already in the existing benchmarks and we should just make sure it does not regress somehow?
> 
> 1. There are existing unit tests for `ray.remote(...)`, `ray.kill(...)` and `ray.list_named_actors(...)`. They still apply here. Testing the RPC stubs directly seems like testing the internal implementations, which Ray is not currently doing. Maybe they are useful in edge cases, but that can be added in future.
> 2. Existing unit tests are mostly CPU bound. The `client: tasks and get batch` benchmark exercises this code path pretty heavily, but don't show a lot of difference. We don't yet have tests with simulated RPC latencies, but we probably should add them in future.
>    Before the change, `client: tasks and get batch per second 1.0 +- 0.0`
>    After the change, `client: tasks and get batch per second 1.03 +- 0.0`
> 
> The result from manually tested more realistic scenario is posted at the beginning.

I feel the microbenchmark is not very useful for some cases. Do you want to add a new one for this work?

File: src/ray/protobuf/ray_client.proto (hunk 21 above)
Comment (mwtian) [reply src/ray/protobuf/ray_client.proto:372]: It is a bit confusing but the handlers of Schedule / ListNamedActors in server.py are called by the streaming Datapath handler to handle the corresponding logic e.g. ClientTask / ClientTaskTicket. Before this change, Get / Put etc are in the same situation, where client - server interacts via Datapath but Get / Put handlers are still exposed. This can be cleaned up in a future PR, after we have a more stable client / server protocol.

**Comment by mwtian** [comment]
> I feel the microbenchmark is not very useful for some cases. Do you want to add a new one for this work?

This work mostly hides round trip latencies between client and server for scheduling. But within a single machine there is not much latency. I can try injecting latencies but it could be a larger change. Maybe in another PR?

**Comment by iycheng** [comment]
> > I feel the microbenchmark is not very useful for some cases. Do you want to add a new one for this work?
> 
> 
> 
> This work mostly hides round trip latencies between client and server for scheduling. But within a single machine there is not much latency. I can try injecting latencies but it could be a larger change. Maybe in another PR?

Thanks for the explanation! I'll treat it as the limitation of micro benchmark. Let's don't do that. We should add a metric later in nightly test for this.

File: src/ray/protobuf/ray_client.proto (hunk 21 above)
Comment (iycheng) [reply src/ray/protobuf/ray_client.proto:372]: Got it. It makes sense!

File: python/ray/includes/unique_ids.pxi (hunk 20 above)
Comment (mwtian) [reply python/ray/includes/unique_ids.pxi:334]: The timeout shouldn't be too long because it is in destructor, and the user doesn't need the ref / object anymore. We can lower the timeout to 5 or 2. Timeout for `ray.get()` is 2s.

When there is a failure to resolve an ID, the ID will not be released so the ref will be kept server-side after the client ref is gone. Object and actor refs are still cleaned up server-side, when client-server shuts down. Failure to resolve an ID most likely happens after a failed connection or server. So this should not be a significant issue. Before this change, this issue can happen in theory if a ClientTaskTicket response is lost, although this is more rare.

**Review by AmeerHajAli — APPROVED** [review]
@mwtian , Thanks for following through with this. It looks very close and people seem to be excited about this. Let me know when you finish addressing the comments and run some manual testing in the product, perhaps @ckw017 can help with the testing (or kill the proxy).

File: python/ray/util/client/worker.py (hunk 3 above)
Comment (ijrsvt) [reply python/ray/util/client/worker.py:449]:
```suggestion
    def call_remote(self, instance, *args, **kwargs) -> List[Future]:
    # TODO(mwtian) Refactor `call_remote`, `call_release` & `call_retain`
    # to have internal prefix
```

File: python/ray/includes/object_ref.pxi (hunk 18 above)
[the remark this answers is not in the document]
Comment (ijrsvt) [reply python/ray/includes/object_ref.pxi:179]: Let me make take a stab at fixing it!

File: python/ray/includes/object_ref.pxi [hunk 22]
```diff
@@ -166,8 +166,19 @@ cdef class ClientObjectRef(ObjectRef):
...
+        if client.ray.is_connected():
+            try:
+                self._wait_for_id(timeout=10)
+            # cython would suppress this exception as well, but it tries to
+            # print out the exception which may crash. Log a simpler message
+            # instead.
+            except Exception:
+                logger.info(
+                    "Exception in ObjectRef is ignored in destructor. "
+                    "To raise the exception, use the object reference before "
```
Comment (ijrsvt) [inline python/ray/includes/object_ref.pxi:178]: Maybe something like:
```
To receive this exception in application code, ray.get() the object 
reference before its destructor is run.
```

**Review by ijrsvt — APPROVED** [review]
This LGTM now! Thanks Mingwei!!!!!

File: python/ray/includes/unique_ids.pxi (hunk 20 above)
Comment (iycheng) [reply python/ray/includes/unique_ids.pxi:334]: I feel this is somewhat still important. It's still a leak and under unstable internet environment it's going to happen a lot. Although it will be released once the client exits, it still leaks in the code and that program will fail.

Btw, how is the reconnect going to work here? Will the failed resolving be called again?

And is it better to retry this and fail the program in the extreme case (for example failed to do cleanup for 5 times)

Or should we make the timeout here to be infinite?

**Review by fishbone — APPROVED** [review]
Nice job!

File: python/ray/includes/unique_ids.pxi (hunk 20 above)
Comment (mwtian) [reply python/ray/includes/unique_ids.pxi]: When the resolving takes too long and fails, if the resolution succeeds in the background with a server response, the ID future will be set but there will be not ref counting. I thought about it for a bit, and it seems there is no issue with using an infinite timeout here. It becomes closer to the behavior before this PR, where client waits synchronously for response from server.

File: python/ray/includes/object_ref.pxi (hunk 22 above)
Comment (mwtian) [reply python/ray/includes/object_ref.pxi]: Thanks for the suggestion!

**Comment by mwtian** [comment]
> @mwtian , Thanks for following through with this. It looks very close and people seem to be excited about this. Let me know when you finish addressing the comments and run some manual testing in the product, perhaps @ckw017 can help with the testing (or kill the proxy).

I believe all comments have been addressed. The change has been tested in product with this [wheel](https://ray-wheels.s3.us-west-2.amazonaws.com/test_wheels/mwtian-local-ref/518e5601d9a3b4e5b960bc90246afda38b56ef37/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl). Reconnection after network disruption works. Also the following code snippet took ~2s with the change instead of ~100s,
```
start = time.perf_counter()
result = [f.remote(i) for i in range(1000)]
ray.get(result)
end = time.perf_counter()
```
This should be ready to be merged now.

**Comment by AmeerHajAli** [comment]
Thanks @mwtian !!!!
